### PR TITLE
fix(agent): watch for changes in configuration files in config directories

### LINKED ISSUE #9985: Telegraf `--watch-config` doesn't support directories
### System info:

Telegraf 1.20 docker image running in a k8s container

### Steps to reproduce:

1. Configure telegraf with the `--watch-config inotify` and `--config-directory` options
2. Have multiple config files inside the config directory
3. Change one of them

### Expected behavior:

I would expect Telegraf to spot that a config file has been changed and restart.

### Actual behavior:

It appears that Telegraf is unaware of the changed file and does not restart.

### Additional info:

https://github.com/influxdata/telegraf/pull/9485
https://github.com/influxdata/telegraf/commit/9508d4f09c13628a8c177a560531a9a0874a1476

### PR DESCRIPTION
**Pull request by iarkhanhelsky**
# Required for all PRs

- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #9985

Another iteration on the fix from #10379 by @conorevans. 

Extracted config.WalkDirectory to collect a list of config files to load configuration from. Reuse the same list later to schedule automatic reloads for changed files.

### CODE REVIEW

**Comment by iarkhanhelsky** [comment]
!signed-cla

**Comment by iarkhanhelsky** [comment]
@reimda @srebhan Could you review this? You had comments on related PR. Thanks!

**Comment by iarkhanhelsky** [comment]
!signed-cla

File: cmd/telegraf/telegraf.go [hunk 1]
```diff
@@ -71,6 +72,7 @@ func (t *Telegraf) Init(pprofErr <-chan error, f Filters, g GlobalFlags, w Windo
 	t.pprofErr = pprofErr
 	t.inputFilters = f.input
 	t.outputFilters = f.output
+	t.configFiles = []string{}
```
Comment (sspaink) [inline cmd/telegraf/telegraf.go:75]:
```suggestion
```

Not necessary as the default value of a slice is a zero-length slice

File: cmd/telegraf/telegraf.go [hunk 2]
```diff
@@ -161,30 +163,31 @@ func (t *Telegraf) watchLocalConfig(signals chan os.Signal, fConfig string) {
...
 			return err
 		}
+		configFiles = append(configFiles, files...)
+	}
+
+	// If no other options are specified, load the config file and run.
+	c := config.NewConfig()
+	c.OutputFilters = t.outputFilters
+	c.InputFilters = t.inputFilters
+	var err error
```
Comment (sspaink) [inline cmd/telegraf/telegraf.go:186]:
```suggestion
```
This `err` seems unused now since it was moved

**Review by sspaink — CHANGES_REQUESTED** [review]
Thank you for working on this! Not sure why the bot doesn't think you've signed the CLA I checked manually and I see that you did. Sorry about that. I've also left some minor suggestions.

**Review by Hipska — APPROVED** [review]

File: cmd/telegraf/telegraf.go (hunk 1 above)
Comment (iarkhanhelsky) [reply cmd/telegraf/telegraf.go]: Updated

File: cmd/telegraf/telegraf.go (hunk 2 above)
Comment (iarkhanhelsky) [reply cmd/telegraf/telegraf.go]: It's still used, moved initialization lower

**Review by sspaink — APPROVED** [review]

**Review by MyaLongmire — APPROVED** [review]

**Pull request merged**